staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1255,23 +1255,28 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 	return heartbeatTime, true
 }
 
-func getEventObject(object runtime.Object) runtime.Object {
+func getMutableObject(object runtime.Object) runtime.Object {
 	if _, ok := object.(runtime.CacheableObject); ok {
 		// It is safe to return without deep-copy, because the underlying
-		// object was already deep-copied during construction.
+		// object will lazily perform deep-copy on the first try to change
+		// any of its fields.
 		return object
 	}
 	return object.DeepCopyObject()
 }
 
+func updateResourceVersion(object runtime.Object, versioner storage.Versioner, resourceVersion uint64) {
+	if err := versioner.UpdateObject(object, resourceVersion); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failure to version api object (%d) %#v: %v", resourceVersion, object, err))
+	}
+}
+
 func updateResourceVersionIfNeeded(object runtime.Object, versioner storage.Versioner, resourceVersion uint64) {
 	if _, ok := object.(*cachingObject); ok {
```

Comment:
do we need the `_IfNeeded` function that special cases *cachingObject if it already handles attempts to set resourceVersion to the existing value as a no-op?

Reply:
It's just slightly faster to check the type than to take a lock:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go#L205-L210

But I don't feel very strongly about this - if you have strong preference towards removing it, I can change it.

Reply:
it's just hard to keep track of all the special cases at the different layers... in the current state of this PR, I think there's only a single call to this now (in the update-makes-object-no-longer-match-filter case). if the performance impact is negligible, it would be easier to reason about if we just always set `updateResourceVersion`

Reply:
changed (added a comment about that too)